docs/dev_cycle.part-04.rst

```diff
@@ -0,0 +1,315 @@
+transaction.
+
+
+.. note::
+
+    Since these are *transactions*, they will change state, and in the case of the ``tester`` chain this state will persist
+    until the test function quits.
+
+
+**[5] Providing arguments**: The donate function in the contract accepts one argument
+
+.. code-block:: solidity
+
+    function donate(uint usd_rate) public payable money_sent {...}
+
+This argument is provided in the test as *Python* donate function:
+
+.. code-block:: python
+
+    donator.transact({'value':650}).donate(380).
+
+Populus gives you a *Python* interface to a bytecode contract. Nice, no?
+
+**[6] Asserts**: We expect the ``donations_total`` to be ``500 + 650 = 1150``, the ``donations_count`` is 2,
+and the ``default_usd_rate`` to match the last update, 380. The test gets the varaibles with ``call``, but in the case
+of a "getter" it doesn't matter, since a "getter" function will not change the contract's state wether it's called with ``transact`` or ``call``.
+
+Run the test:
+
+.. code-block:: bash
+
+    $ py.test --disable-pytest-warnings
+
+    platform linux -- Python 3.5.2, pytest-3.1.3, py-1.4.34, pluggy-0.4.0
+    rootdir: /home/mary/projects/donations, inifile:
+    plugins: populus-1.8.0, hypothesis-3.14.0
+    collected 2 items
+
+    tests/test_donator.py ..
+
+    ===================== 2 passed, 10 warnings in 0.58 seconds =============
+
+Voila. The two tests pass.
+
+
+Test Calculations
+-----------------
+
+
+The next one will test the ETH/USD calculations:
+
+.. code-block:: shell
+
+    $ nano tests/test_donator.py
+
+Add the following test to the bottom of the file:
+
+.. code-block:: python
+
+    def test_usd_calculation(chain):
+
+        ONE_ETH_IN_WEI = 10**18  # 1 ETH == 1,000,000,000,000,000,000 Wei
+
+        donator, deploy_tx_hash = chain.provider.get_or_deploy_contract('Donator')
+        donator.transact({'value':ONE_ETH_IN_WEI}).donate(400)
+        donator.transact({'value':(2 * ONE_ETH_IN_WEI)}).donate(500)
+        donations_usd = donator.call().donations_usd()
+
+        # donated 1 ETH in  $400 per ETH = $400
+        # donated 2 ETH in $500 per ETH = 2 * $500 = $1,000
+        # total $ value donated = $400 + $1,000 = $1,400
+        assert donations_usd == 1400
+
+The test sends donations worth of 3 Ether. Reminder: by default, all contract functions
+and contract interactions are handled in *Wei*.
+
+In 1 Ether we have 10^18 Wei (see the `Ether units denominations <http://ethdocs.org/en/latest/ether.html>`_)
+
+The test runs two transactions: note the ``transact`` function, which will change the contract state and balance
+on the blockchain. We use the ``tester`` chain, so the state is reset on each test run.
+
+**First transaction**
+
+.. code-block:: python
+
+    donator.transact({'value':ONE_ETH_IN_WEI}).donate(400)
+
+Donate Wei worth of 1 Ether, where the effective ETH/USD rate is $400. That is, $400 per Ether,
```

Comment:
Rather than using `$400` which is both above current market value, and could be interpreted as a price projection, can we use a less real number like $5 which is way below market value and more of a *made-up* number?

Reply:
ok